firedrake/matrix_free/preconditioners.py

```diff
@@ -0,0 +1,311 @@
+
+        """
+        pass
+
+
+class AssembledPC(PCBase):
+    """A matrix-free PC that assembles the operator.
+
+    Internally this makes a PETSc PC object that can be controlled by
+    options using the extra options prefix ``assembled_``.
+    """
+    def initialize(self, pc):
+        from firedrake.assemble import assemble
+
+        _, P = pc.getOperators()
+
+        context = P.getPythonContext()
+
+        # It only makes sense to preconditioner/invert a diagonal
+        # block in general.  That's all we're going to allow.
+        if not context.on_diag:
+            raise ValueError("Only makes sense to invert diagonal block")
+
+        self.P = assemble(context.a, bcs=context.row_bcs,
+                          form_compiler_parameters=context.fc_params, nest=False)
+        self.P.force_evaluation()
+
+        # Transfer nullspace over
+        Pmat = self.P.petscmat
+        Pmat.setNullSpace(P.getNullSpace())
+        Pmat.setTransposeNullSpace(P.getTransposeNullSpace())
+
+        prefix = pc.getOptionsPrefix()
+
+        # Internally, we just set up a PC object that the user can configure
+        # however from the PETSc command line.  Since PC allows the user to specify
+        # a KSP, we can do iterative by -assembled_pc_type ksp.
+        pc = PETSc.PC().create()
+        pc.setOptionsPrefix(prefix+"assembled_")
+        pc.setOperators(Pmat, Pmat)
+        pc.setUp()
+        pc.setFromOptions()
+        self.pc = pc
+
+    def update(self, pc):
+        from firedrake import assemble
+        P = self.P
+        P = assemble(P.a, tensor=P, bcs=P.bcs)
+        P.force_evaluation()
+
+    def apply(self, pc, x, y):
+        self.pc.apply(x, y)
+
+    def applyTranspose(self, pc, x, y):
+        self.pc.applyTranspose(x, y)
+
+    def view(self, pc, viewer=None):
+        super(AssembledPC, self).view(pc, viewer)
+        viewer.printfASCII("PC to apply inverse\n")
+        viewer.pushASCIITab()
+        self.pc.view(viewer)
+        viewer.popASCIITab()
+
+
+class MassInvPC(PCBase):
+
+    """A matrix free operator that inverts the mass matrix in the provided space.
+
+    Internally this creates a PETSc KSP object that can be controlled
+    by options using the extra options prefix ``Mp_``.
+    """
+    def initialize(self, pc):
+        from firedrake import TrialFunction, TestFunction, dx, assemble, inner
+        prefix = pc.getOptionsPrefix()
+
+        # we assume P has things stuffed inside of it
+        _, P = pc.getOperators()
+        context = P.getPythonContext()
+
+        test, trial = context.a.arguments()
+
+        if test.function_space() != trial.function_space():
+            raise ValueError("MassInvPC only makes sense if test and trial space are the same")
+
+        V = test.function_space()
+
+        u = TrialFunction(V)
+        v = TestFunction(V)
+        # Handle vector and tensor-valued spaces.
+        a = inner(u, v)*dx
```

Comment:
That is correct, however you use the viscosity weighting for the variable viscosity case (hence "low-contrast").  I'll note that this is not good for high viscosity contrasts, see, for example https://arxiv.org/abs/1607.03936